bika/lims/api.py

```diff
@@ -464,11 +461,11 @@ def get_object_by_uid(uid, default=_marker):
 
     # we defined the portal object UID to be '0'::
     if uid == '0':
-        return get_portal()
+        return get_portal(context=context)
 
     # we try to find the object with both catalogs
-    pc = get_portal_catalog()
-    uc = get_tool("uid_catalog")
+    pc = get_portal_catalog(context=context)
+    uc = get_tool("uid_catalog", context=context)
```

Comment:
I tried various ways to localise changes to get_tool but none worked. This is the error I get http://docs.plone.org/develop/plone.api/docs/api/exceptions.html#plone.api.exc.CannotGetPortalError

Reply:
Ok, let's try if we can handle that somehow in `get_tool`/`get_portal` only, so that we do not have to change the signature of all other functions.

Please add this to the `api.get_tool`:

```python
def get_tool(name, context=None, default=_marker):
    # Try first with the context
    if context is not None:
        try:
            context = get_object(context)
            return getToolByName(context, name)
        except (BikaLIMSError, AttributeError) as e:
            # https://github.com/senaite/bika.lims/issues/396
            logger.warn("get_tool::getToolByName({}, '{}') failed: {} "
                        "-> falling back to plone.api.portal.get_tool('{}')"
                        .format(repr(context), name, repr(e), name))
            return get_tool(name, default=default)

    # Try with the plone api
    try:
        return ploneapi.portal.get_tool(name)
    except InvalidParameterError:
        if default is not _marker:
            return default
        fail("No tool named '%s' found." % name)

    # Handle weird Plone API Error
    # https://docs.plone.org/develop/plone.api/docs/api/exceptions.html#plone.api.exc.CannotGetPortalError
    except CannotGetPortalError:
        if default is not _marker:
            return default
        return get_tool(name, context=get_portal(), default=None)
```

Then we have to ensure that `get_portal` does not fail:

```python
def get_app():
        import Zope2
        return Zope2.app()

def get_portal():
    """Get the portal object
    :returns: Portal object
    """
    try
        return ploneapi.portal.getSite()
    except CannotGetPortalError:
        app = get_app()
        sites = app.objectValues("Plone Site")
        if len(sites) != 1:
            return None
        return sites[0]
```

Reply:
This is just something that went into my mind, please feel free to optimize and change. If we go that way I would suggest a new PR for the API including an updated doctest.
Name of the PR should be then "Handle CannotGetPortalError in API" or so.

Reply:
Maybe the `get_portal` function can be refined with the request, e.g.:
```python
def get_portal():
    """Get the portal object
    :returns: Portal object
    """
    try
        return ploneapi.portal.getSite()
    except CannotGetPortalError:
        app = get_app()

        request = get_request()
        portal_id = "Try to get the portal id out of the request... "
        if portal_id in app.objectIds():
            return app[portal_id]
        sites = app.objectValues("Plone Site")
        if len(sites) != 1:
            return None
        return sites[0]
```

Reply:
I tested the above possible solutions but to no avail. The error occurs when publishing in _analyses_data which is called from the digester. It fails in my new unit conversion code but at the start of _analyses_data method are some getToolByName calls that also fail when I change them to get_tool. I'm not saying my solution is pretty but it's not the worst hack ever.

Reply:
Not at all, please don't understand it as such.
However, this solution fixes the symptoms, but not the reason. It is something that should not happen and I'm reluctant that the API loses it's simplicity for developers by the need to understand why this context needs to get passed everywhere.

From my gut feeling, this can be all traced back to the Report Digester. I would rather see that one go instead of adding code to workaround that weirdness – it will haunt us forever if we do not tackle that properly now.

Reply:
Please could you prove that this only happens when you get the Analyses from the Digester. Try to disable the Digester completely and check if the error still occurs. If it does not, we will remove Digester in another PR

Reply:
I merged the ARDigester methods in ARPublishView class and call _ar_data directly and the error does not occur.